src/polygon.rs

```diff
@@ -0,0 +1,110 @@
+    }
+    return points;
+}
+
+pub struct PolygonData {
+    pub polygon_stamped_msg: Option<rosrust_msg::geometry_msgs::PolygonStamped>,
+    pub lines_in_static_frame: Option<Vec<Line>>,
+    _color: Color,
+    _tf_listener: Arc<rustros_tf::TfListener>,
+    _static_frame: String,
+}
+
+pub struct PolygonListener {
+    _data: Arc<RwLock<PolygonData>>,
+    _subscriber: rosrust::Subscriber,
+}
+
+impl PolygonData {
+    pub fn update(&mut self) {
+        self.lines_in_static_frame = None;
+        if let Some(polygon) = &self.polygon_stamped_msg {
+            let transform = self._tf_listener.clone().lookup_transform(
+                &self._static_frame.clone(),
+                &polygon.header.frame_id,
+                polygon.header.stamp,
+            );
+
+            match &transform {
+                Ok(transform) => {
+                    let mut lines: Vec<Line> = Vec::new();
+                    let polygon_transform = ros_transform_to_isometry(&transform.transform);
+                    let mut previous_point: Option<Point3<f64>> = None;
+                    for pt in read_points(&polygon.polygon) {
+                        let point_in_static_frame = polygon_transform.transform_point(&pt);
+                        if let Some(pp) = previous_point {
+                            lines.push(Line {
+                                x1: pp.x,
+                                y1: pp.y,
+                                x2: point_in_static_frame.x,
+                                y2: point_in_static_frame.y,
+                                color: self._color.clone(),
+                            });
+                        }
+                        previous_point = Some(point_in_static_frame);
+                    }
+
+                    self.lines_in_static_frame = Some(lines);
+                }
+                Err(_e) => {}
+            };
+        }
+    }
+
+    pub fn get_lines(&self) -> Vec<Line> {
+        if let Some(lines) = &self.lines_in_static_frame {
+            return lines.to_vec();
+        }
+        return vec![];
+    }
+}
+impl PolygonListener {
+    pub fn new(
+        config: ListenerConfigColor,
+        tf_listener: Arc<rustros_tf::TfListener>,
+        static_frame: String,
+    ) -> PolygonListener {
+        let data = Arc::new(RwLock::new(PolygonData {
+            polygon_stamped_msg: None,
+            lines_in_static_frame: None,
+            _tf_listener: tf_listener,
+            _static_frame: static_frame,
+            _color: config.color.to_tui(),
+        }));
+
+        let cloned_data = data.clone();
+        let sub = rosrust::subscribe(
+            &config.topic,
+            1,
+            move |msg: rosrust_msg::geometry_msgs::PolygonStamped| {
+                let mut unlocked_data = cloned_data.write().unwrap();
+                unlocked_data.polygon_stamped_msg = Some(msg);
+                unlocked_data.update();
```

Comment:
@carzum  this update is translating the polygon to the correct position.

Currently we do this every time we receive a new polygon but do not work great on latched topic.

Moving this to the `get_lines` method it is also not ideal because it will slow down the thread that we use for  the visual rendering. Should we anyway move it there and in the future think of adding a separate timer with the task of transforming all the subscribed messages?
?

Reply:
the draw thread is the one constraining performance at the moment. Having the transform in there could make the application more unresponsive. Regarding latched topics: We have the same issue with other listeners, so far this is not bugging me too much.